### PR TITLE
Allow custom tokenization

### PR DESCRIPTION
**Pull request by kornelski**
As a workaround for #3, this allows user to use custom split implementation.

### CODE REVIEW

File: src/rake.rs [hunk 1]
```diff
@@ -26,7 +26,16 @@ impl Rake {
...
-        let phrases = self.phrases(text);
+        let phrases = self.phrases(PUNC_RE.split(text));
+        let word_scores = self.word_scores(&phrases);
+        self.candidate_keywords(&phrases, word_scores)
+    }
+
+    /// Runs RAKE algorithm on chunks of text and returns a vector of keywords.
+    /// The returned vector is sorted by score (from greater to less).
+    #[inline]
+    pub fn run_sentences<'a>(&self, sentences: impl IntoIterator<Item = &'a str>) -> Vec<KeywordScore> {
```
Comment (yaa110) [inline src/rake.rs:37]: 1. I think it should be renamed to `run_by_iterator` or something like that.
2. this function might return wrong result by passing an invalid iterator. So, it should be mentioned in the documentation of this function.

File: src/rake.rs [hunk 2]
```diff
@@ -26,7 +26,16 @@ impl Rake {
     /// Runs RAKE algorithm on `text` and returns a vector of keywords.
     /// The returned vector is sorted by score (from greater to less).
     pub fn run(&self, text: &str) -> Vec<KeywordScore> {
```
Comment (yaa110) [inline src/rake.rs:28]: please add `#[inline]` for this function, too.

File: src/rake.rs [hunk 3]
```diff
@@ -77,24 +86,26 @@ impl Rake {
         word_score
     }
 
-    fn phrases<'a>(&'a self, text: &'a str) -> Vec<Vec<&'a str>> {
-        let mut phrases = Vec::new();
-        PUNC_RE.split(text).filter(|s| !s.is_empty()).for_each(|s| {
+    fn phrases<'a>(&self, phrases_iter: impl IntoIterator<Item = &'a str>) -> Vec<Vec<&'a str>> {
+        let phrases_iter = phrases_iter.into_iter();
+        let mut phrases = Vec::with_capacity(2 * phrases_iter.size_hint().0);
+        for s in phrases_iter.filter(|s| !s.is_empty()) {
```
Comment (yaa110) [inline src/rake.rs:92]: why did you replace `for_each`with `for` loop?

File: src/rake.rs [hunk 4]
```diff
@@ -77,24 +86,26 @@ impl Rake {
         word_score
     }
 
-    fn phrases<'a>(&'a self, text: &'a str) -> Vec<Vec<&'a str>> {
-        let mut phrases = Vec::new();
-        PUNC_RE.split(text).filter(|s| !s.is_empty()).for_each(|s| {
+    fn phrases<'a>(&self, phrases_iter: impl IntoIterator<Item = &'a str>) -> Vec<Vec<&'a str>> {
+        let phrases_iter = phrases_iter.into_iter();
+        let mut phrases = Vec::with_capacity(2 * phrases_iter.size_hint().0);
```
Comment (yaa110) [inline src/rake.rs:91]: isn't it better to use `collect` in the following chain rather than pushing elements into this vector?

**Review by yaa110 — CHANGES_REQUESTED** [review]

File: src/rake.rs (hunk 1 above)
Comment (kornelski) [reply src/rake.rs:37]: I think naming it after a language construct is not correct. It should use domain-specific name. Same as you use `let phrases =`, not `let variable = `.

File: src/rake.rs (hunk 2 above)
Comment (kornelski) [reply src/rake.rs:28]: It doesn't use monomorphisation, so it's unlikely to benefit from that.

File: src/rake.rs (hunk 1 above)
Comment (yaa110) [reply src/rake.rs:37]: I agree, but `run_sentences` is somewhat ambiguous.

File: src/rake.rs (hunk 3 above)
Comment (kornelski) [reply src/rake.rs:92]: It seems clearer, given that it's not a self-contained iterator, but rather the loop mutates the outer variable.

File: src/rake.rs (hunk 4 above)
Comment (kornelski) [reply src/rake.rs:91]: Not every iteration gives an element, so turning it into an iterator usable with collect would be tricky.

File: src/rake.rs (hunk 1 above)
Comment (kornelski) [reply src/rake.rs:37]: phrases? chunks? fragments? I'm not sure what's the technically best terminology here. The implementation used to split on punctuation which is roughly equivalent to sentences?

**Comment by kornelski** [comment]
It's not clear to me what needs to be done here, so I'm leaving it to you.

**Comment by dbxnr** [comment]
What's the status with this?

**Comment by kornelski** [comment]
Most of it has been merged into 3.0

**Pull request closed without merging**